drv/transceivers-server/src/main.rs

```diff
@@ -112,6 +158,178 @@ impl ServerImpl {
+
+        // Wait for the I2C transaction to complete
+        loop {
+            self.transceivers
+                .get_i2c_status_and_read_buffer(port, &mut out)?;
+            if out[0] & Reg::QSFP::PORT0_STATUS::BUSY == 0 {
+                break;
+            }
+        }
+
+        if out[0] & Reg::QSFP::PORT0_STATUS::ERROR == 0 {
+            match out[1] {
+                0x1E => Ok(ManagementInterface::Cmis),
+                0x0D | 0x11 => Ok(ManagementInterface::Sff8636),
+                i => Ok(ManagementInterface::Unknown(i)),
+            }
+        } else {
+            // TODO: how should we handle this?
+            // Right now, we'll retry on the next pass through the loop.
+            Err(FpgaError::ImplError(0))
+        }
+    }
+
+    fn update_thermal_loop(&mut self, status: ModulesStatus) {
+        let now = sys_get_timer().now;
+
+        for i in 0..self.thermal_models.len() {
+            let port = LogicalPort(i as u8);
+            let mask = 1 << i;
+            let powered =
+                (status.present & mask) != 0 && (status.power_good & mask) != 0;
+
+            // A wild transceiver just appeared!  Read it to decide whether it's
+            // using SFF-8636 or CMIS.
+            if powered && self.thermal_models[i].is_none() {
+                match self.get_transceiver_interface(port) {
+                    Ok(interface) => {
+                        ringbuf_entry!(Trace::GotInterface(i, interface));
+                        // TODO: this is made up
+                        self.thermal_models[i] = Some(ThermalModel {
+                            interface,
+                            model: task_thermal_api::ThermalProperties {
+                                target_temperature: Celsius(65.0),
+                                critical_temperature: Celsius(70.0),
+                                power_down_temperature: Celsius(80.0),
+                                temperature_slew_deg_per_sec: 0.5,
+                            },
+                        });
+                    }
+                    Err(e) => {
+                        // Not much we can do here if reading failed
+                        ringbuf_entry!(Trace::TemperatureReadError(i, e));
+                    }
+                }
+            } else if !powered && self.thermal_models[i].is_some() {
+                // This transceiver went away; remove it from the thermal loop
+                if let Err(e) = self.thermal_api.remove_dynamic_input(i) {
+                    ringbuf_entry!(Trace::ThermalError(i, e));
+                }
+
+                ringbuf_entry!(Trace::UnpluggedModule(i));
+                self.thermal_models[i] = None;
+            }
+        }
+
+        for (i, m) in self.thermal_models.iter().enumerate() {
+            let port = LogicalPort(i as u8);
+            let m = match m {
+                Some(m) => m,
+                None => continue,
+            };
```

Comment:
Might be too clever, but I think you could squish this down into the iterator:

```rust
        for (port, m) in self
            .thermal_models
            .iter()
            .enumerate()
            .filter_map(|(i, m)| m.map(|m| (LogicalPort(i as u8), m)))
        {
```

It's not any shorter, but arguably might be slightly more obvious that the for loop is only operating on a subset of `self.thermal_models`?